File: src/g1_camera.cpp
#include "g1_camera.hpp"

#include <chrono>
#include <stdexcept>

#include <viam/sdk/log/logging.hpp>

namespace viam_unitree {

G1Camera::G1Camera(const viam::sdk::Dependencies& deps, const viam::sdk::ResourceConfig& cfg)
    : Camera(cfg.name()) {
    auto attrs = cfg.attributes();
    std::string network_interface = "eth0";

    auto it = attrs.find("network_interface");
    if (it != attrs.end()) {
        if (auto* val = it->second.get<std::string>()) {
            network_interface = *val;
        }
    }

    VIAM_SDK_LOG(info) << "Initializing G1Camera with network interface: " << network_interface;

    unitree::robot::ChannelFactory::Instance()->Init(0, network_interface);
    video_client_.Init();

    VIAM_SDK_LOG(info) << "G1Camera initialized successfully";
}

std::vector<std::string> G1Camera::validate(const viam::sdk::ResourceConfig& cfg) {
    return {};
}

viam::sdk::Camera::image_collection G1Camera::get_images(
    std::vector<std::string> filter_source_names,
    const viam::sdk::ProtoStruct& extra) {
    std::vector<uint8_t> image_data;
    int32_t rc = video_client_.GetImageSample(image_data);
    if (rc != 0) {
        throw std::runtime_error("GetImageSample failed with rc=" + std::to_string(rc));
    }

    viam::sdk::Camera::raw_image img;
    img.mime_type = "image/jpeg";
    img.bytes = std::vector<unsigned char>(image_data.begin(), image_data.end());
    img.source_name = name();

    viam::sdk::Camera::image_collection collection;
    collection.images = {std::move(img)};
    collection.metadata.captured_at =
        std::chrono::time_point_cast<std::chrono::nanoseconds>(std::chrono::system_clock::now());

    return collection;
}

viam::sdk::Camera::point_cloud G1Camera::get_point_cloud(std::string mime_type,
                                                         const viam::sdk::ProtoStruct& extra) {
    throw std::runtime_error("point cloud not supported");
}

viam::sdk::Camera::properties G1Camera::get_properties() {
    viam::sdk::Camera::properties props;
    props.supports_pcd = false;
    props.intrinsic_parameters = {};
    props.distortion_parameters = {};
    props.mime_types = {"image/jpeg"};
    props.frame_rate = 30.0f;
    return props;
}

viam::sdk::ProtoStruct G1Camera::get_status() {
    return {};
}

viam::sdk::ProtoStruct G1Camera::do_command(const viam::sdk::ProtoStruct& command) {
    return {};
}

std::vector<viam::sdk::GeometryConfig> G1Camera::get_geometries(
    const viam::sdk::ProtoStruct& extra) {
    return {};
}

}  // namespace viam_unitree

File: g1_camera.go
package main

import (
	"context"
	"fmt"
	"image"
	"time"

	"go.viam.com/rdk/components/camera"
	"go.viam.com/rdk/gostream"
	"go.viam.com/rdk/logging"
	"go.viam.com/rdk/pointcloud"
	"go.viam.com/rdk/resource"
	"go.viam.com/rdk/rimage"
	"go.viam.com/rdk/rimage/transform"
	rdkutils "go.viam.com/rdk/utils"
)

var g1CameraModel = resource.NewModel("erh", "viam-unitree", "g1-camera")

type G1CameraConfig struct {
	NetworkInterface string `json:"network_interface"`
}

func (c *G1CameraConfig) Validate(path string) ([]string, error) {
	return nil, nil
}

func init() {
	resource.RegisterComponent(camera.API, g1CameraModel, resource.Registration[camera.Camera, *G1CameraConfig]{
		Constructor: newG1Camera,
	})
}

type g1Camera struct {
	resource.Named
	resource.AlwaysRebuild

	logger logging.Logger
	video  *VideoClient
}

func newG1Camera(ctx context.Context, deps resource.Dependencies, conf resource.Config, logger logging.Logger) (camera.Camera, error) {
	cfg, err := resource.NativeConfig[*G1CameraConfig](conf)
	if err != nil {
		return nil, err
	}

	networkInterface := "eth0"
	if cfg.NetworkInterface != "" {
		networkInterface = cfg.NetworkInterface
	}

	logger.Infof("Initializing G1Camera with network interface: %s", networkInterface)

	if err := InitChannel(0, networkInterface); err != nil {
		return nil, fmt.Errorf("channel init: %w", err)
	}

	video, err := NewVideoClient()
	if err != nil {
		return nil, fmt.Errorf("video client create: %w", err)
	}
	if err := video.Init(); err != nil {
		video.Close()
		return nil, fmt.Errorf("video client init: %w", err)
	}

	logger.Info("G1Camera initialized successfully")

	return &g1Camera{
		Named:  conf.ResourceName().AsNamed(),
		logger: logger,
		video:  video,
	}, nil
}

func (c *g1Camera) Image(ctx context.Context, mimeType string, extra map[string]interface{}) ([]byte, camera.ImageMetadata, error) {
	jpegData, err := c.video.GetImage()
	if err != nil {
		return nil, camera.ImageMetadata{}, err
	}
	return jpegData, camera.ImageMetadata{MimeType: rdkutils.MimeTypeJPEG}, nil
}

func (c *g1Camera) Images(ctx context.Context) ([]camera.NamedImage, resource.ResponseMetadata, error) {
	jpegData, err := c.video.GetImage()
	if err != nil {
		return nil, resource.ResponseMetadata{}, err
	}

	img, err := rimage.DecodeImage(ctx, jpegData, rdkutils.MimeTypeJPEG)
	if err != nil {
		return nil, resource.ResponseMetadata{}, fmt.Errorf("decode jpeg: %w", err)
	}

	return []camera.NamedImage{
			{Image: img, SourceName: c.Name().ShortName()},
		}, resource.ResponseMetadata{
			CapturedAt: time.Now(),
		}, nil
}

func (c *g1Camera) Stream(ctx context.Context, errHandlers ...gostream.ErrorHandler) (gostream.VideoStream, error) {
	return gostream.NewEmbeddedVideoStreamFromReader(gostream.VideoReaderFunc(func(ctx context.Context) (image.Image, func(), error) {
		jpegData, err := c.video.GetImage()
		if err != nil {
			return nil, nil, err
		}
		img, err := rimage.DecodeImage(ctx, jpegData, rdkutils.MimeTypeJPEG)
		if err != nil {
			return nil, nil, fmt.Errorf("decode jpeg: %w", err)
		}
		return img, func() {}, nil
	})), nil
}

func (c *g1Camera) NextPointCloud(ctx context.Context) (pointcloud.PointCloud, error) {
	return nil, fmt.Errorf("point cloud not supported")
}

func (c *g1Camera) Properties(ctx context.Context) (camera.Properties, error) {
	return camera.Properties{
		SupportsPCD: false,
		MimeTypes:   []string{rdkutils.MimeTypeJPEG},
		FrameRate:   30.0,
		ImageType:   camera.ColorStream,
	}, nil
}

func (c *g1Camera) Projector(ctx context.Context) (transform.Projector, error) {
	return nil, transform.NewNoIntrinsicsError("")
}

func (c *g1Camera) DoCommand(ctx context.Context, cmd map[string]interface{}) (map[string]interface{}, error) {
	return map[string]interface{}{}, nil
}

func (c *g1Camera) Close(ctx context.Context) error {
	c.video.Close()
	return nil
}
